lib/palettes_base.cpp

```diff
@@ -0,0 +1,337 @@
+PaletteRegistry::PaletteRegistry(const vector<Palette>& paletteRefs)
+{
+   for (const Palette& palette : paletteRefs)
+   {
+      if (IsNameUnique(palette.name))
+      {
+         palettes.push_back(as_unique<Palette>(palette));
+      }
+   }
+}
+
+void PaletteRegistry::AddPalette(Palette& palette)
+{
+   if (IsNameUnique(palette.name))
+   {
+      palettes.push_back(as_unique<Palette>(palette));
+   }
+}
+
+void PaletteRegistry::AddPalette(const string& name)
+{
+   if (IsNameUnique(name))
+   {
+      palettes.push_back(as_unique<Palette>(name));
+   }
+}
+
+bool PaletteRegistry::IsNameUnique(const string& name) const
+{
+   // palette name is unique || container is empty
+   if (GetIndexByName(name) == -1 || palettes.empty())
+   {
+      return true;
+   }
+   else
+   {
+      cout << "Palette with name: '" << name << "' already exists in registry.";
+      return false;
+   }
+}
+
+Palette* PaletteRegistry::get(int index) const
+{
+   if (0 <= index && index <= NumPalettes()-1)
+   {
+      return palettes[index].get();
+   }
+   cout << "Palette (index = " << index+1 << ") out of range. Available palettes:"
+        << endl;
+   this->PrintSummary();
+   return palettes.back().get();
+}
+
+Palette* PaletteRegistry::get(const string& name) const
+{
+   int idx = GetIndexByName(name);
+   if (idx != -1)
+   {
+      return palettes[idx].get();
+   }
+   cout << "Palette (name = " << name << ") not found. Available palettes:" <<
+        endl;
+   this->PrintSummary();
+   return palettes.back().get();
+}
+
+void PaletteRegistry::PrintSummary(ostream& os) const
+{
+   for (int i = 0; i < NumPalettes(); i++)
+   {
+      os << setw(3) << i+1 << ") "
+         << left << setw(12) << get(i)->name << right;
+      if ((i+1)%5 == 0)
+      {
+         os << endl;
+      }
+   }
+   os << endl;
+}
+
+void PaletteRegistry::PrintAll(ostream& os) const
+{
+   for (int i = 0; i < NumPalettes(); i++)
+   {
+      get(i)->Print(os);
+   }
+}
+
+void PaletteRegistry::Load(const string& palette_filename)
+{
+   ifstream pfile(palette_filename);
+   if (!pfile)
+   {
+      cout << "Could not open palette file: " << palette_filename << endl;
+      return;
+   }
+   string word, palname, channeltype;
+   int idx = -1;
+
+   // read initializing commands
+   while (1)
+   {
+      pfile >> ws;
+      if (!pfile.good())
+      {
+         break;
+      }
+      if (pfile.peek() == '#')
+      {
+         getline(pfile, word);
+         continue;
+      }
+      pfile >> word;
+      if (word == "palette")
+      {
+         pfile >> palname >> channeltype;
+         idx = GetIndexByName(palname);
+         if (idx == -1)
+         {
+            AddPalette(palname);
+            idx = GetIndexByName(palname);
+            cout << "Reading palette: (" << idx+1 << ") " << palname << endl;
+         }
+         else
+         {
+            cout << "Error reading palette: " << palname
+                 << ". Palette with same name already exists." << endl;
```

Comment:
I do not know, does this need to be an error stopping the loading? Maybe you have two large collections, which partially overlap, and you do not want to go through them manually to remove duplicates. I would prefer a warning and skipping or rewriting the palette.

Reply:
Thats fair - we don't query by name yet but it will be nice in `pyglvis` and in glvis `evil` mode 😈 . I like the idea of overwriting because that seems like the more common scenario

Reply:
That overwriting looks simple here, just empty the existing one and the rest should work :wink: